apps/meteor/app/reactions/server/setReaction.ts

```diff
@@ -51,40 +38,62 @@ async function setReaction(room: IRoom, user: IUser, message: IMessage, reaction
 	// 	return;
 	// }
 
-	const userAlreadyReacted =
-		message.reactions &&
-		Boolean(message.reactions[reaction]) &&
-		message.reactions[reaction].usernames.indexOf(user.username as string) !== -1;
-	// When shouldReact was not informed, toggle the reaction.
+	const reactedUsernamesForEmoji = message.reactions?.[reaction]?.usernames ?? [];
+
+	const userIndexInReactionList = reactedUsernamesForEmoji.indexOf(user.username as string);
+
+	const userAlreadyReacted = userIndexInReactionList !== -1;
+
 	if (shouldReact === undefined) {
 		shouldReact = !userAlreadyReacted;
 	}
 
-	if (userAlreadyReacted === shouldReact) {
+	if (shouldReact === userAlreadyReacted) {
 		return;
 	}
 
-	let isReacted;
+	let didReact;
 
 	if (userAlreadyReacted) {
+		// 1. remove username from reaction list
+		// 2. if reaction list for that emoji was already 1, remove the reaction entry (for frontend)
+		// 3. if reaction list now is empty, remove 'reactions' field altogether
+
 		const oldMessage = JSON.parse(JSON.stringify(message));
-		removeUserReaction(message, reaction, user.username as string);
-		if (_.isEmpty(message.reactions)) {
-			delete message.reactions;
-			if (isTheLastMessage(room, message)) {
-				await Rooms.unsetReactionsInLastMessage(room._id);
+
+		if (reactedUsernamesForEmoji?.length === 1) {
+			// we can just pull the whole 'emoji' field
+			// @ts-ignore
+			if (message.reactions && Object.keys(message.reactions).length === 0) {
```

Comment:
```suggestion
			if (message.reactions && Object.keys(message.reactions).length === 1) {
```
Isn't that the case for when the user removed the last reaction of a message (and no reaction is left after that)? Since this happens before the last reaction is removed, this should be `1`, or am I missing something?

Reply:
you're right .. i had the key delete above this statement before, but then moved